Source/Details/ASCollectionLayoutState.h

```diff
@@ -0,0 +1,54 @@
+//
+//  ASCollectionLayoutState.h
+//  AsyncDisplayKit
+//
+//  Created by Huy Nguyen on 9/3/17.
+//  Copyright © 2017 Facebook. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+#import <AsyncDisplayKit/ASBaseDefines.h>
+
+@class ASElementMap, ASCollectionElement, ASLayout;
+
+NS_ASSUME_NONNULL_BEGIN
+
+AS_SUBCLASSING_RESTRICTED
+@interface ASCollectionLayoutState : NSObject
+
+/// The element map used to calculate this object
+@property (nonatomic, weak, readonly) ASElementMap *elementMap;
```

Comment:
My main reason for using `weak` here and in `ASCollectionLayoutContext` is to make sure only `ASDataController` holds strong reference to the map and thus it is released as soon as `ASDataController` has a new data or it's deallocated. That being said, having a strong reference here ensures that the state is valid during its lifetime and I updated the code in favor of this. The take away is that now `visibleMap` is retained not only by `ASDataController` but also by other objects, so we need to be careful to not leak it.